Support DRY type-annotated @rules. (#8330)
pantsbuild/pants: src/python/pants/engine/README.md

## API

The end user API for the engine is based on the registration of `@rule`s, which are functions

> or coroutines with statically declared parameter types (also known as `Param` types) and return type

Author: As I read it the grouping suggested by 1 is: "coroutines with statically declared (parameter types and return type)" whereas in 2 it's "coroutines with (statically declared parameter types) and (a return type)". The latter grouping is not what we want the reader to hear. This may be in the ear of the beholder though.

(also known as the `Product` type). A Pants (plugin) developer can write and install additional

> `@rule`s to extend the functionality of Pants.

Author: Out of scope.

### Products and Params

The engine executes your `@rule`s in order to (recursively) compute a `Product` of the requested
type for a set of `Param`s. This recursive type search leads to a loosely coupled (and yet still 
statically checked) form of dependency injection.

When an `@rule` runs, it requires a set of `Param`s that the engine has determined are needed to
compute its transitive `@rule` dependencies. So although an `@rule` might not have a particular
`Param` type in its signature, it might depend on another `@rule` that does need that `Param`, and
would thus need that `Param` in order to run. To see which `Params` the engine needs to run each
`@rule`, refer to the `Visualization` section below.

Any hashable type with useful equality may be used as a `Param`, and additional `Params` can be
provided to an `@rule`'s dependencies via `Get` requests (see below). Each `Param` value in a set of
`Params` is unique by type, so if `@rule`s recursively introduce a particular `Param` type, there
will still only be one value for that type in each `@rule`, but it will change as you move deeper

> into the dependency graph.

Reviewer: This is a correct description of what *should* be happening, but there is a bug about this that I believe is still unfixed specifically regarding performing a `yield Get()` with the same subject/product pair twice in the same `@rule` body (where pants fails at runtime). I hadn't yet made an issue about this because I was trying to fix it really quickly, but if I can't do that today I'll make an issue to track this. I believe @illicitonion and @gshuflin have run into this.

Author: OK. Thanks in advance for filing. I won't change language here per my out of scope mantra though.

#### Example

As a very simple example, you might register the following `@rule` that can compute a `str`
`Product` given a single `int` argument.

```python
@rule
def int_to_str(value: int) -> str:
  return str(value)
```

> Here the only requirements for the `@rule` are:

Reviewer: ```suggestion
Here, the only requirements for the `@rule` are:
```
subjective/optional

Author: Fixed.

> 1. The decorated function is a free, top-level function; ie: methods cannot be `@rule`s.

Reviewer: ```suggestion
1. The decorated function is a free, top-level function; i.e.: methods cannot be `@rule`s.
```
subjective/optional

Author: Fixed.

> 2. The decorated function must be fully type annotated.

Reviewer: I would like to keep this simple, unequivocal language, and it seems that if we do, we would also want to actually deprecate the old `@rule` syntax (without annotations) in this PR as well? Or are we referring to either syntax here?

Author: I think its fine that the language is simple without the deprecation in the same PR. In docs we steer folks to use annotations by not mentioning the soon to be deprecated syntax. No need to mention the old syntax since the new syntax covers everything needed by a user.

Reviewer: Hm, ok. I might decide to avoid deprecating `datatype` in the same PR as #8181 then to unblock it!

> 3. As mentioned above, the types of the function `Param`s must be hashable and have useful equality.

Reviewer 2: It's very subtle, but: a `@rule`s arguments are not necessarily exactly the `Params` it requires. See the second paragraph of the "Products and Params" section. In a few other places this document refers to "argument selectors", but we don't have a more formal distinction than that.

Author: Ok - I'll strike. In general I was trying to remove mention of engine internals, but this is a mized document already. Much of its language is imo noisy for rule writers, but important for engine hackers. I'll probably be circling in soon to break up the document along those lines, or otherwise extract the relevant bits and expand so @rule writers can get to business and not get confused unless they need to dive deep.

Reviewer: That sounds like a really good idea! I've been linking to the engine README to others to try to explain "what it is", and I think the potential split you've described would allow us to have a document that could serve that purpose, while also keeping around the notes for engine hackers elsewhere.

When the engine encounters this `@rule` while compiling the rule graph for `str`-producing-`@rule`s,
it will next go hunting for the dependency `@rule` that can produce an `int` using the fewest number
of `Params`. For example, if there was an `@rule` that could produce an `int` without consuming any
`Params` at all (ie, a singleton), then that `@rule` would always be chosen first. If all `@rule`s
to produce `int`s required at least one `Param`, then the engine would next see whether the input
`Params` contained an `int`, or whether there were any `@rule`s that required only one `Param`, then
two `Params`, and so on.

In cases where this search detects any ambiguity (generally because there are two or more `@rule`s
that can provide the same product with the same number of parameters), rule graph compilation will

> fail with a useful error message.

Reviewer: Here is where the colloquialism "rulesets" might come in handy, because the ambiguity can occur not just from having aliasing for specific `@rule` product/output pairs, but as a property of a whole subgraph. This distinction hasn't come up a ton yet in practice yet except in the native backend due to the (currently) small number of `@rule`s in the codebase. My first stab at trying to allude to that:
```suggestion
fail with a useful error message. Note that sometimes ambiguity can also occur as a result of sets of `@rule`s which produce two separate paths through the rule graph with overlapping parameters and output product, without any specific `@rule` or pair of `@rule`s causing the ambiguity.
```

Author: I'm going to defer on out of scope changes. I only:
+ Fixed 100 col line length limit.
+ Fixed @rule syntax guidance.

Reviewer: That sounds great! Thank you!

### Datatypes

In practical use, builtin types like `str` or `int` do not provide enough information to
disambiguate between various types of data in `@rule` signatures, so declaring small `datatype` 

> definitions to provide a unique and descriptive type is highly recommended:

Reviewer: (thanks for updating the paragraph filling in this doc!)

Author: Ditto

Author: I did nothing of the sort that I can tell, I just line-wrapped.

Assigning a specific type to a field can be somewhat unidiomatic in Python, and may be unexpected or
unnatural to use. However, regardless of whether the object is created directly with type-checked
fields or whether it's produced from a set of rules by the engine's dependency injection, it is
extremely useful to formalize the assumptions made about the value of an object into a specific
type, even if the type just wraps a single field. The `datatype()` function makes it simple and
efficient to apply that strategy.

> ### Gets and RootRules

Reviewer: ```suggestion
### `Get`s and `RootRule`s
```
subjective/optional

Author: Ditto
